sky/provision/oci/instance.py

```diff
@@ -0,0 +1,455 @@
+                    source_details=oci_adaptor.oci.core.models.
+                    InstanceSourceViaImageDetails(
+                        source_type='image',
+                        image_id=node_config['ImageId'],
+                        boot_volume_size_in_gbs=node_config['BootVolumeSize'],
+                        boot_volume_vpus_per_gb=int(
+                            node_config['BootVolumePerf']),
+                    ),
+                    create_vnic_details=oci_adaptor.oci.core.models.
+                    CreateVnicDetails(
+                        assign_public_ip=True,
+                        subnet_id=vcn,
+                    ),
+                    shape_config=machine_shape_config,
+                    preemptible_instance_config=preempitible_config,
+                ))
+
+            new_inst = launch_instance_response.data
+            if not head_instance_id:
+                head_instance_id = new_inst.id
+                logger.debug(f'New head node: {head_instance_id}')
+
+            created_instances.append({
+                'inst_id': new_inst.id,
+                'name': new_inst.display_name,
+                'ad': new_inst.availability_domain,
+                'compartment': new_inst.compartment_id,
+                'status': new_inst.lifecycle_state,
+                'oci_tags': new_inst.freeform_tags,
+            })
+        # end for loop
+    # end if to_start_count > 0:...
+
+    for inst in (resume_instances + created_instances):
+        logger.debug(f'Provisioning for node {inst["name"]}')
+        _wait_until_status(region, inst['inst_id'], 'RUNNING')
+        logger.debug(f'Instance {inst["name"]} is RUNNING.')
+
+    total_time = round(time.time() * 1000) - start_time
+    logger.debug('Total time elapsed: {0} milli-seconds.'.format(total_time))
+
+    assert head_instance_id is not None, head_instance_id
+
+    return common.ProvisionRecord(
+        provider_name='oci',
+        region=region,
+        zone=None,
+        cluster_name=cluster_name_on_cloud,
+        head_instance_id=head_instance_id,
+        created_instance_ids=[n['inst_id'] for n in created_instances],
+        resumed_instance_ids=[n['inst_id'] for n in resume_instances],
+    )
+
+
+@query_utils.debug_enabled(logger)
+def stop_instances(
+    cluster_name_on_cloud: str,
+    provider_config: Dict[str, Any],
+    worker_only: bool = False,
+) -> None:
+    """Stop running instances."""
+    assert provider_config is not None, (cluster_name_on_cloud, provider_config)
+
+    region = provider_config['region']
+    tag_filters = {constants.TAG_RAY_CLUSTER_NAME: cluster_name_on_cloud}
+    if worker_only:
+        tag_filters[constants.TAG_RAY_NODE_KIND] = 'worker'
+
+    nodes = _get_filtered_nodes(region, tag_filters)
+    for node in nodes:
+        oci_adaptor.get_core_client(
+            region, oci_utils.oci_config.get_profile()).instance_action(
+                instance_id=node['inst_id'], action='STOP')
```

Comment:
Could we put those core api calls to the query_helper like `query_helper.terminate_instances_by_tags`?

Reply:
Put all core api related calls to the query_helper.